src/dragonfly_haiku_emscripten.rs

```diff
@@ -0,0 +1,87 @@
+// Copyright 2018 Developers of the Rand project.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// https://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or https://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Implementation for DragonFly / Haiku
+use Error;
+use std::fs::File;
+use std::io::Read;
+use std::num::NonZeroU32;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::os::unix::io::{RawFd, AsRawFd, FromRawFd};
+use std::{thread, mem};
+
+// replace with AtomicI32 on stabilization and MSRV bump
+static RNG_FD: AtomicUsize = AtomicUsize::new(-1i32 as usize);
+// replace with AtomicU8 on stabilization and MSRV bump
+static RNG_STATE: AtomicUsize = AtomicUsize::new(0);
+
+const STATE_INIT_ONGOING: usize = 1 << 0;
+const STATE_INIT_DONE: usize = 1 << 1;
+
+pub fn getrandom_inner(dest: &mut [u8]) -> Result<(), Error> {
+    let fd = RNG_FD.load(Ordering::Acquire) as i32;
+    if fd >= 0 {
+        return use_fd(fd, dest);
+    }
+    let fd = loop {
+        let state = RNG_STATE.fetch_or(STATE_INIT_ONGOING, Ordering::AcqRel);
+        if state & STATE_INIT_DONE != 0 {
+            // initialization is complete, use fd from atomic
+            break RNG_FD.load(Ordering::Acquire) as RawFd;
+        } else if state & STATE_INIT_ONGOING == 0 {
+            // start initialization and return resulting fd
+            match init_fd() {
+                Ok(fd) => {
+                    RNG_STATE.store(STATE_INIT_DONE, Ordering::Release);
+                    RNG_FD.store(fd as usize, Ordering::Release);
```

Comment:
I believe we need to store the FD *before* the STATE (it is `Ordering::Release` on the last write which matters).

Reply:
Hm, you are right, but maybe we should even change ordering to `SeqCst` here? Either way I would like someone to specifically review use of atomics in this PR.